urllib3/filepost.py

```diff
@@ -56,39 +53,165 @@ def iter_fields(fields):
-            data = str(data)  # Backwards compatibility
+    def next(self):
+        return six.next(self.iterator)
 
-        if isinstance(data, six.text_type):
-            writer(body).write(data)
-        else:
-            body.write(data)
+    def __next__(self):
+        return self.next()
+
+    def read(self, size=None):
+        data = self.leftover
+        count = len(self.leftover)
+        try:
+            while count < size:
+                chunk = self.next()
+                data += chunk
+                count += len(chunk)
+        except StopIteration:
+            pass
 
-        body.write(b'\r\n')
+        if count > size:
+            self.leftover = data[size:]
 
-    body.write(b('--%s--\r\n' % (boundary)))
+        return data[:size]
 
-    content_type = str('multipart/form-data; boundary=%s' % boundary)
 
-    return body.getvalue(), content_type
+class MultipartEncoderGenerator(object):
+    """
+    Generator yielding chunk-by-chunk streaming data from fields, with proper
+    headers and boundary separators along the way. This is useful for streaming
+    large files as iterators without loading the entire data body into memory.
+
+    ``fields`` is a dictionary where the parameter name is the key and the value
+    is either a (filename, data) tuple or just data.
+
+    The data can be a unicode string, an iterator producing strings, or a file-like
+    object. File-like objects are read ``chunk_size`` bytes at a time.
+
+    If no ``boundary`` is specified then a random one is used.
+    """
+    def __init__(self, fields, boundary=None, chunk_size=8192):
+        self.fields = [field for field in iter_field_objects(fields)]
+        self.chunk_size = chunk_size
+        self.boundary = boundary or choose_boundary()
+
+    def get_content_type(self):
+        return 'multipart/form-data; boundary=%s' % self.boundary
+
+    def __len__(self):
+        """
+        Figure out the expected body size by iterating over the fields as if they
+        contained empty files, while accumulating the value file sizes as
+        efficiently as we can.
+        """
+        empty_fields = []
+        size = 0
+        for field in self.fields:
+            name = field._name
+            data = field.data
+            filename = field._filename
+
+            if filename is not None:
+                empty_fields.append((name, (filename, '')))
+            else:
+                empty_fields.append((name, ''))
+
+            if hasattr(data, '__len__'):
+                size += len(encode(data))
+            elif isinstance(data, six.integer_types):
+                size += len(str(data))
+            elif hasattr(data, 'seek'):
+                size += file_size(data)
+            elif hasattr(data, 'read'):
+                size += len(encode(data.read()))  # This is undesired
+            elif hasattr(data, '__iter__'):
+                size += sum(len(encode(chunk)) for chunk in data)  # This is also undesired
```

Comment:
That's exactly what they do, hence the undesired behavior comments ;) These were a part of the old stream that this work is based on so I copied them over. I was thinking we could hold the data inside of a buffer after getting the length but I wanted to get it looked at before I made that change.